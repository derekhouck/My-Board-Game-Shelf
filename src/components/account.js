import React from 'react';
import { connect } from 'react-redux';
import requiresLogin from './requires-login';

export function Account(props) {
  return (
    <section>
      <h2>Your Account</h2>
      <ul>
        <li>
          Name: {props.currentUser.name}
        </li>
        <li>
          Username: {props.currentUser.username}
        </li>
      </ul>
    </section>
  );
}

const mapStateToProps = state => ({
  currentUser: state.auth.currentUser
});

export default requiresLogin()(connect(mapStateToProps)(Account));