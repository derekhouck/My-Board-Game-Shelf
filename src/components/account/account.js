import React from 'react';
import { connect } from 'react-redux';
import requiresLogin from '../requires-login';
import { toggleDeleting } from '../../actions/users';

import Button from '../button';
import EditAccount from './edit-account';
import DeleteAccount from './delete-account';
import ChangePassword from './change-password';

export class Account extends React.Component {
  state = {
    editing: false,
    passwordChange: false
  }

  setEditing(editing) {
    this.setState({ editing });
  }

  setPasswordChange(passwordChange) {
    this.setState({ passwordChange });
  }

  render() {
    const { currentUser, deleting, dispatch } = this.props;
    if (this.state.editing) {
      return <EditAccount
        resetEditing={() => this.setEditing(false)}
      />;
    } else if (this.state.passwordChange) {
      return <ChangePassword
        resetPasswordChange={() => this.setPasswordChange(false)}
      />;
    } else if (deleting) {
      return <DeleteAccount />;
    } else {
      return (
        <section>
          <h1>Your Account</h1>
          <ul>
            <li>
              Name: {currentUser.name}
            </li>
            <li>
              Email: {currentUser.email}
            </li>
            <li>
              Username: {currentUser.username}
            </li>
          </ul>
          <Button
            label="Edit Account"
            onClick={() => this.setEditing(true)}
          />
          <div>
            <h2>Change Your Password</h2>
            <Button
              label="Change Password"
              onClick={() => this.setPasswordChange(true)}
              secondary
            />
          </div>
          <div>
            <h2>Delete Your Account</h2>
            <p>
              <strong>Warning: This action cannot be undone!</strong>
            </p>
            <Button
              primary
              className="delete-account-button"
              onClick={() => dispatch(toggleDeleting())}
              label="Delete Account"
            />
          </div>
        </section>
      );
    }
  }
}

const mapStateToProps = state => ({
  currentUser: state.auth.currentUser,
  deleting: state.users.deleting
});

export default requiresLogin()(connect(mapStateToProps)(Account));