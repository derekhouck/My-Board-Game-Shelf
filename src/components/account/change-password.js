import React from 'react';
import { connect } from 'react-redux';
import { focus, reduxForm, Field } from 'redux-form';
import { isTrimmed, length, matches, nonEmpty, required } from '../../validators';

import Button from '../button';
import Input from '../atoms/input';
import { editUser } from '../../actions/users';

const passwordLength = length({ min: 8, max: 72 });
const matchesPassword = matches('password');

export class ChangePassword extends React.Component {
  onSubmit(values) {
    const { password } = values;
    const updateData = {
      id: this.props.currentUser.id,
      password
    };

    return this.props.dispatch(editUser(updateData));
  }

  render() {
    return (
      <form
        className="change-password-form"
        onSubmit={this.props.handleSubmit(values => this.onSubmit(values))}
      >
        <fieldset>
          <legend>Change Your Password</legend>
          <Field
            component={Input}
            type="password"
            name="password"
            id="password"
            label="New Password"
            validate={[required, isTrimmed, passwordLength]}
          />
          <Field
            component={Input}
            type="password"
            name="passwordConfirm"
            id="passwordConfirm"
            label="Confirm new password"
            validate={[required, matchesPassword, nonEmpty]}
          />
        </fieldset>
        <div className="form-actions btn-group">
          <Button
            disabled={this.props.pristine || this.props.submitting}
            type="submit"
            label="Change Password"
            primary
          />
          <Button
            label="Cancel"
            onClick={() => this.props.resetPasswordChange()}
            secondary
          />
        </div>
      </form>
    );
  }
}

const mapStateToProps = state => ({
  currentUser: state.auth.currentUser
});

export default reduxForm({
  form: 'change-password-form',
  onSubmitFail: (errors, dispatch) => dispatch(focus('change-password-form', Object.keys(errors)[0]))
})(connect(mapStateToProps)(ChangePassword));