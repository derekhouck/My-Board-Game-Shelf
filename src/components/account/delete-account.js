import React from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, focus } from 'redux-form';
import { Link } from 'react-router-dom';
import { toggleDeleting, deleteUser } from '../../actions/users';
import { required, nonEmpty, matches } from '../../validators';

import Button from '../atoms/button';
import Input from '../atoms/input';

const matchesUsername = matches('username');


export class DeleteAccount extends React.Component {
  componentDidMount() {
    const initData = {
      "username": this.props.currentUser.username
    }
    this.props.initialize(initData);
  }

  componentWillUnmount() {
    this.props.dispatch(toggleDeleting());
  }

  onSubmit(values) {
    this.props.dispatch(deleteUser(this.props.currentUser));
  }

  render() {
    const username = this.props.currentUser.username;

    return (
      <section>
        <h2>Are you sure you want to delete your account?</h2>
        <p>This will delete all the games in your collection. This action cannot be undone.</p>
        <p>If you still want to proceed, type your username, <strong>{username}</strong>, into the text box below.</p>
        <form
          className="delete-account-form"
          onSubmit={this.props.handleSubmit(values => this.onSubmit(values))}
        >
          <Field
            component="input"
            type="hidden"
            name="username"
          />
          <Field
            component={Input}
            type="text"
            name="confirmUsername"
            id="confirmUsername"
            validate={[required, nonEmpty, matchesUsername]}
          />
          <Button
            primary
            type="submit"
            label="Delete My Account"
          />
        </form>
        <Link to="/dashboard">
          <Button
            label="I've changed my mind"
          />
        </Link>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  currentUser: state.auth.currentUser
});

export default reduxForm({
  form: 'delete-account',
  onSubmitFail: (errors, dispatch) => dispatch(focus('registration', Object.keys(errors)[0]))
})(connect(mapStateToProps)(DeleteAccount));