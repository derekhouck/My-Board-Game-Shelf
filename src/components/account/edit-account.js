import React from 'react';
import { connect } from 'react-redux';
import { focus, reduxForm, Field } from 'redux-form';
import { isTrimmed, required } from '../../validators';
import { editUser } from '../../actions/users';

import Button from '../button';
import Input from '../input';

export class EditAccount extends React.Component {
  componentDidMount() {
    const initData = {
      "name": this.props.currentUser.name,
      "username": this.props.currentUser.username
    };
    this.props.initialize(initData);
  }

  onSubmit(values) {
    const { name, username } = values;
    const updateData = {
      id: this.props.currentUser.id,
      name,
      username
    };

    return this.props.dispatch(editUser(updateData));
  }

  render() {
    return (
      <form
        className="edit-account-form"
        onSubmit={
          this.props.handleSubmit(values => this.onSubmit(values))
        }
      >
        <fieldset>
          <legend>Edit Your Account</legend>
          <Field
            component={Input}
            type="text"
            name="name"
            label="Name"
          />
          <Field
            component={Input}
            type="text"
            name="username"
            label="Username"
            validate={[required, isTrimmed]}
          />
        </fieldset>
        <div className="form-actions btn-group">
          <Button
            disabled={this.props.pristine || this.props.submitting}
            type="submit"
            label="Save Edits"
            primary
          />
          <Button
            label="Cancel"
            secondary
            onClick={() => this.props.resetEditing()}
          />
        </div>
      </form>
    );
  }
}

const mapStateToProps = state => ({
  currentUser: state.auth.currentUser
});

export default reduxForm({
  form: 'edit-account-form',
  onSubmitFail: (errors, dispatch) => dispatch(focus('edit-account-form', Object.keys(errors)[0]))
})(connect(mapStateToProps)(EditAccount));