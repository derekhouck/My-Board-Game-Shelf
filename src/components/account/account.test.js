import React from 'react';
import { shallow } from 'enzyme';
import { Account } from './account';

describe('<Account />', function () {
  it('renders without crashing', () => {
    shallow(<Account currentUser={{}} />);
  });

  it('calls the toggleDeleting action when the "Delete Account" button is clicked', function () {
    const dispatch = jest.fn();
    const mockToggleDeletingAction = { type: 'TOGGLE_DELETING' };
    const wrapper = shallow(<Account currentUser={{}} dispatch={dispatch} />);
    wrapper.find('.delete-account-button').simulate('click');
    expect(dispatch).toHaveBeenCalledWith(mockToggleDeletingAction);
  });
});
