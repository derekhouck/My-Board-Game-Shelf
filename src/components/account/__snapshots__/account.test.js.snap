// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account /> renders without crashing 1`] = `ShallowWrapper {}`;
